src/host/readDataCooked.cpp

```diff
@@ -797,453 +817,477 @@ void COOKED_READ_DATA::_transitionState(State state) noexcept
-    const til::point contentOrigin{
-        (viewportSize.width - contentSize.width) / 2 + viewportOrigin.x,
-        (viewportSize.height - contentSize.height) / 2 + viewportOrigin.y,
-    };
-    const til::rect contentRect{
-        contentOrigin,
-        contentSize,
+    return {
+        .offset = static_cast<size_t>(it - beg),
+        .column = column,
     };
-    const auto backupRect = Viewport::FromExclusive({
-        contentRect.left - 1,
-        contentRect.top - 1,
-        contentRect.right + 1,
-        contentRect.bottom + 1,
-    });
-
-    auto& popup = _popups.emplace_back(kind, contentRect, backupRect, std::vector<CHAR_INFO>{ widthSizeT * heightSizeT });
-
-    // Create a backup of the TextBuffer parts we're scribbling over.
-    // We need to flush the buffer to ensure we capture the latest contents.
-    // NOTE: This may theoretically modify popup.backupRect (practically unlikely).
-    _flushBuffer();
-    THROW_IF_FAILED(ServiceLocator::LocateGlobals().api->ReadConsoleOutputWImpl(_screenInfo, popup.backup, backupRect, popup.backupRect));
-
-    // Draw the border around our content and fill it with whitespace to prepare it for future usage.
-    {
-        const auto attributes = _screenInfo.GetPopupAttributes();
+}
 
-        RowWriteState state{
-            .columnBegin = contentRect.left - 1,
-            .columnLimit = contentRect.right + 1,
-        };
+void COOKED_READ_DATA::_appendCUP(std::wstring& output, til::point pos)
+{
+    fmt::format_to(std::back_inserter(output), FMT_COMPILE(L"\x1b[{};{}H"), pos.y + 1, pos.x + 1);
+}
 
-        // top line ┌───┐
-        std::wstring buffer;
-        buffer.assign(widthSizeT, L'─');
-        buffer.front() = L'┌';
-        buffer.back() = L'┐';
-        state.text = buffer;
-        textBuffer.Replace(contentRect.top - 1, attributes, state);
-
-        // bottom line └───┘
-        buffer.front() = L'└';
-        buffer.back() = L'┘';
-        state.text = buffer;
-        textBuffer.Replace(contentRect.bottom, attributes, state);
-
-        // middle lines │   │
-        buffer.assign(widthSizeT, L' ');
-        buffer.front() = L'│';
-        buffer.back() = L'│';
-        for (til::CoordType y = contentRect.top; y < contentRect.bottom; ++y)
-        {
-            state.text = buffer;
-            textBuffer.Replace(y, attributes, state);
-        }
-    }
+void COOKED_READ_DATA::_appendPopupAttr(std::wstring& output) const
+{
+    VtIo::FormatAttributes(output, _screenInfo.GetPopupAttributes().GetLegacyAttributes());
+}
+
+void COOKED_READ_DATA::_popupPush(const PopupKind kind)
+try
+{
+    auto& popup = _popups.emplace_back(kind);
+    _dirty = true;
 
     switch (kind)
     {
-    case PopupKind::CopyToChar:
-        _popupDrawPrompt(popup, ID_CONSOLE_MSGCMDLINEF2);
-        break;
-    case PopupKind::CopyFromChar:
-        _popupDrawPrompt(popup, ID_CONSOLE_MSGCMDLINEF4);
-        break;

```

Comment:
Why no more `CopyToChar` and `CopyFromChar`?

Reply:
All of the drawing logic moved over to "redisplay". I did this because with VT we can't directly manipulate the buffer anymore, so we need to be smart and correctly calculate the positions of the text and the popups before we even write them out.